core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -1227,6 +1227,62 @@ public static SqlCall stripSeparator(SqlCall call) {
             .createSqlType(SqlTypeName.VARCHAR, typePrecision);
       };
 
+  /**
+   * Type-inference strategy for String concatenation.
+   * Result is binary type when arguments contain binary type; otherwise return varchar type.
+   * For example,
+   *
+   * <p>concat(cast('a' as varchar(1)), cast('b' as varchar(1)),cast('c' as varchar(1)))
+   * returns varchar(3).
+   *
+   * <p>concat(x'61', x'62', x'63')
+   * returns binary(3).
+   *
+   * <p>concat(x'61', 'b', x'63')
+   * returns binary(3).
+   *
+   */
+  public static final SqlReturnTypeInference STRING_BYTESTRING_PRECISION =
+      opBinding -> {
+        boolean hasPrecisionNotSpecifiedOperand = false;
+        boolean precisionOverflow = false;
+        int typePrecision;
+        long amount = 0;
+        List<RelDataType> operandTypes = opBinding.collectOperandTypes();
+        boolean containsByteStringType = containsByteStringType(operandTypes);
+        final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        final RelDataTypeSystem typeSystem = typeFactory.getTypeSystem();
+        for (RelDataType operandType : operandTypes) {
+          int operandPrecision = operandType.getPrecision();
+          amount = (long) operandPrecision + amount;
```

Comment:
is this precision correct for a mix of strings and byte arrays?
what if a character is unicode and spans multiple bytes?
please add a test for this case too.